new view for a specific version of a resource

mockersf opened the pull request:
related to #3734 

New view that display all informations related to a specific version of a resource. It's a copy / paste of the resource view, but it will display only one version and it's expanded by default:
<img width="1043" alt="resource version" src="https://user-images.githubusercontent.com/8672791/58387466-ce447080-800e-11e9-8cb3-2065562d11d6.png">

This screen can be reached by clicking the `#` in the resource view:
<img width="1043" alt="resource" src="https://user-images.githubusercontent.com/8672791/58387467-dd2b2300-800e-11e9-8da1-712e21beafd8.png">

Or the `#` in the build view on a  resource:
<img width="1043" alt="build" src="https://user-images.githubusercontent.com/8672791/58387472-f0d68980-800e-11e9-8ce2-fb1f04154889.png">

For me, the main utility is to be able to go from build -> resource version -> other builds that are using the same version of this resource

Other information that could be interesting to display on this view:
- the logs that produce this version, but I think that would be a bigger change to get the logs 
- anything else ? I'm open to suggestions.

If the view is OK I think there should be more work done for code deduplication between resource view and resource version view, but I didn't want to go into that before discussing what goes where

pivotal-jwinters commented:
@vito what about using the version hash? We use that as the identifier elsewhere, and then we won't have to deal with weird characters in the url.

vito commented:
@pivotal-jwinters I think hashes would end up even being confusing when - for example if I'm trying to link to commit `1985b7128198358dde2390ec69ede303ff5962eb` the URL would be something like `e739c0e0fe1853d7a909e5657dceb78e` instead. :sweat_smile: We also don't expose the hashes in the UI or API - they're just an internal detail to make uniqueness easier in the database schema atm.

Since versions are currently expected to be fairly regular (simple strings) I think we can probably get away with having readable URLs. We may run into a similar challenge in the future for spaces since they'll also be routing based on 'config fragments' (assuming concourse/rfcs#24) like `branch:abcdef`.

mockersf commented:
1/ the "`#`" looks ok to me but it's not great, it can change

2/ yes, it's copy pasted and can be removed

3/ as the version is a free json object, it can get quite complicated, contains illegal characters in a URL, be very long (in a toy resource I once put a 5MB version just to see if it worked...)... and not be very readable once url encoded
I found this library that tries to do something clever to help on this subject : https://github.com/Sage/jsurl
sample json: (can't be used directly in a url, has to be encoded)
```
{"name":"John Doe","age":42,"children":["Mary","Bill"]}
```
url encoded json: (safe to use in a url but can get very long)
```
%7B%22name%22%3A%22John%20Doe%22%2C%22age%22%3A42%2C%22children%22%3A%5B%22Mary%22%2C%22Bill%22%5D%7D
```
encoded with jsurl: (looks readable and mostly safe, but made for query params, path params have additional contraints to url encode)
```
~(name~'John*20Doe~age~42~children~(~'Mary~'Bill))
```
base 64: (which is the safest from a url point of view, but not readable at all)
```
eyJuYW1lIjoiSm9obiBEb2UiLCJhZ2UiOjQyLCJjaGlsZHJlbiI6WyJNYXJ5IiwiQmlsbCJdfQ==
```
custom encoding: (a lot of special cases to think about. For example, what if a value contains a ","?)
```
name:John%20Doe,age:42,children:mary,bill
```

vito commented:
We ended up wanting another form of querying by version fields for `fly pin-resource-version`: https://github.com/concourse/concourse/pull/3131#issuecomment-459874740 - so I think we'll have to bite the bullet at some point and figure out how these should be passed to the API.

I've opened an issue for it here: https://github.com/concourse/concourse/issues/3932

It might be going a bit too far to block this PR on it, though. :thinking: Unless you're willing to take a swing at that, @mockersf, I'm fine with merging this with IDs in the URL for now.

pivotal-jamie-klassen commented:
> the version is a free json object
@mockersf hopefully a helpful clarification: it has to deserialize to a `map[string]string` https://github.com/concourse/concourse/blob/master/atc/resource_types.go#L7.

Lindsayauchin commented:
@mockersf  

**Feedback on the UX of Adding the # to the resource page:** 
I think we need to take a look at the UX implications of creating a button to link out in that area on the UI of the Resource page beside the pin toggle and the disable/enable toggle. 

In the Resource page, a link out to a new view could make sense on the **version** but, that currently expands and collapses the version meta data and inputs to/outputs.  If  we went down that path of adding a link out there to a new view we would have to create two clickable hotspots on that container. This could confuse users in the short term from the current expected behaviour of the expand and collapse reveal.

Just to get a better understanding of your current work flow. Your goal is to see a list of all builds that use a version of a resource from the build you are currently viewing. 

The pain you are currently experiencing is by navigating back to the pipeline view selecting that resource, going to the resource page then parsing out the version from that list, expanding that version to then view the builds this inputs to so you can navigate to those builds. 

The solutions presented could use some further exploration but, can you confirm my synopsis of your use case?

mockersf commented:
@Lindsayauchin yes you have it right

In the resource page, the link is on another button to not be on expand / collapse toggle. That's not very pretty, but I think the one on the build screen is worse, as there are no button there so I added the link in the expanse / collapse toggle space which is confusing

@vito I removed the check bar

pivotal-jamie-klassen commented:
Could work if the build steps had an arrow that you click to expand/collapse and then the version value itself was a link?

mockersf commented:
I'm not convinced with an array, as there is already a downward arrow on the build screen to mark a get step.

Maybe a small triangle that would turn when expanded?

pivotal-jamie-klassen commented:
@mockersf - I had to mentally `:s/array/arrow/` there, haha. Turning triangle is solid, or a plus/minus toggle would be distinct as well. Definitely worth getting @Lindsayauchin's input on this one.

Lindsayauchin commented:
@mockersf what do you think about adding the entry point for this new view as a link from a button that we put on the build step of that version-> expand that resource version -> button on the meta data area links out to the other builds that are using the same version of this resource? 

![Screen Shot 2019-06-03 at 2 11 47 PM](https://user-images.githubusercontent.com/27746614/58824209-91c5d580-8609-11e9-9f7c-419f592228b5.png)

vito commented:
IMO there's value in having a version-specific page anyway, and having the builds link directly to the versions used by it would make the pipeline feel more cohesive.

jama22 commented:
@vito do you mean to say that this could lead to causality? mind == blown

vito commented:
@jama22 Yeah, that'd be a good page for it in the future. 👨💥

Lindsayauchin commented:
@mockersf @vito 

What if the resource version could be reach via link on the version string with an old school underline style? We would just need to turn off the expand collapse hotspot in that area of the resource container div. 

![Screen Shot 2019-06-12 at 2 34 40 PM](https://user-images.githubusercontent.com/27746614/59377329-faa30100-8d1f-11e9-9272-0be9d1ac4bef.png)

Update the resource page and hide some of the unnecessary elements like the "checking successfully' and the refresh.  

![Screen Shot 2019-06-12 at 2 42 49 PM](https://user-images.githubusercontent.com/27746614/59377535-6b4a1d80-8d20-11e9-8ec4-299f9c8c7943.png)

xtremerui commented:
Hi @mockersf  just a status check that are you still working on this PR?

jomsie commented:
Hi @mockersf! 

We've recently revamped our PR policy, as seen here: https://blog.concourse-ci.org/nobody-wants-a-stale-bot/#improving-pull-request-engagement.

I'm going to go ahead and close this PR for the time being, let me know if you'd like to continue working on it!

Also @matthewpereira I'm not sure if you've seen this PR, but if you have any opinions on it, feel free posting here!
